src/performance.js

```diff
@@ -256,15 +286,51 @@ export class Performance {
    *   also be provided to indicate when to record the tick at.
    * @param {function()=} opt_flush callback function that is called
    *   when we are ready for the ticks to be forwarded to an endpoint.
+   * @param {function(!Object<string,string>)=} opt_setParams callback
+   *   function that is invoked to append additional url parameters
+   *   on the next `flush`.
    */
-  setTickFunction(tick, opt_flush) {
+  setTickFunction(tick, opt_flush, opt_setParams) {
     this.tick_ = tick;
     this.flush_ = opt_flush;
-    this.flushQueuedTicks_();
-    // We need to call flush right away in case `setTickFunction` is called
-    // later than the amp codebase had invoked the performance services'
-    // `flush` method to forward ticks.
-    this.flush();
+    this.setParams_ = opt_setParams;
+    this.whenTickFunctionSetResolve_();
+  }
+
+  /**
+   * Resolved when the tick function is set externally.
+   * @return {!Promise}
+   */
+  whenTickFunctionSet() {
+    return this.whenTickFunctionSetPromise_;
+  }
+
+  /**
+   * Calls setParams with relevant document information.
+   * @return {!Promise}
+   * @private
+   */
+  setDocumentInfoParams_() {
+    return all([
+      this.whenTickFunctionSet(),
+      this.whenViewportLayoutComplete_()
+    ]).then(() => {
+      const params = Object.create(null);
+      params['canonicalUrl'] = documentInfoFor(this.win).canonicalUrl;
+
+      this.resources_.get().forEach((r) => {
+        const name = r.element.tagName.toLowerCase();
+        if (!params[name]) {
+          params[name] = 1;
```

Comment:
Wanna do the `amp-ad` `type` in this round of changes as well? Basically the same deal as for tagNames.